kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,895 @@
+fn is_ordered(requested_ordering: &[ReorderIndex]) -> bool {
+    if requested_ordering.is_empty() {
+        return true;
+    }
+    // we have >=1 element. check that the first element is ordered
+    if !requested_ordering[0].is_ordered() {
+        return false;
     }
+    // now check that all elements are ordered wrt. each other, and are internally ordered
+    requested_ordering
+        .windows(2)
+        .all(|ri| (ri[0].index() < ri[1].index()) && ri[1].is_ordered())
 }
 
+// we use this as a placeholder for an array and its associated field. We can fill in a Vec of None
+// of this type and then set elements of the Vec to Some(FieldArrayOpt) for each column
+type FieldArrayOpt = Option<(Arc<ArrowField>, Arc<dyn ArrowArray>)>;
+
 /// Reorder a RecordBatch to match `requested_ordering`. For each non-zero value in
 /// `requested_ordering`, the column at that index will be added in order to returned batch
-pub(crate) fn reorder_record_batch(
-    input_data: RecordBatch,
-    requested_ordering: &[usize],
-) -> DeltaResult<RecordBatch> {
-    if requested_ordering.windows(2).all(|is| is[0] < is[1]) {
-        // indicies is already sorted, meaning we requested in the order that the columns were
+pub(crate) fn reorder_struct_array(
+    input_data: StructArray,
+    requested_ordering: &[ReorderIndex],
+) -> DeltaResult<StructArray> {
+    if is_ordered(requested_ordering) {
+        // indices is already sorted, meaning we requested in the order that the columns were
         // stored in the parquet
         Ok(input_data)
     } else {
         // requested an order different from the parquet, reorder
-        let input_schema = input_data.schema();
-        let mut fields = Vec::with_capacity(requested_ordering.len());
-        let reordered_columns = requested_ordering
-            .iter()
-            .map(|index| {
-                fields.push(input_schema.field(*index).clone());
-                input_data.column(*index).clone() // cheap Arc clone
-            })
+        debug!("Have requested reorder {requested_ordering:#?} on {input_data:?}");
+        let num_rows = input_data.len();
+        let (input_fields, mut input_cols, null_buffer) = input_data.into_parts();
+        let mut final_fields_cols: Vec<FieldArrayOpt> = std::iter::repeat_with(|| None)
+            .take(requested_ordering.len())
+            .collect();
+        for (parquet_position, reorder_index) in requested_ordering.iter().enumerate() {
+            // for each item, reorder_index.index() tells us where to put it, and its position in
+            // requested_ordering tells us where it is in the parquet data
+            match reorder_index {
+                ReorderIndex::Child { index, children } => {
+                    let mut placeholder: Arc<dyn ArrowArray> =
+                        Arc::new(StructArray::new_empty_fields(0, None));
+                    std::mem::swap(&mut input_cols[parquet_position], &mut placeholder);
```

Comment:
This could potentially reorder every field in a given row yes? Would performance be a concern here? it's not unusual to see thousands of columns in a table. Maybe something like the arc swap crate (which says it's more performant, not sure how true that is) to consider here, but I suppose there is no easy in-place style op to do here that would eliminate the placeholder style?

Reply:
Yes it could, but I don't think the performance of swapping two pointers is the issue. What this does is:
- swap two pointers (should be fast)
- call `into_data().into()` on the swapped out pointer (this has some cost, we convert to `ArrayData` and then right back to `StructArray`. I'd like to find a cheaper way to do this, see my comment about wanting a cheaper way to do this as it's actually just a cast)
- re-order the children (this is required, just a recursive call)
- build a new metadata field to match the new order (fairly fast, and has to happen if we've reordered)

Reordering children is just moving around `Arc`s, so it's pretty cheap. So really the only bit that I think could be faster is `into_data().into()` bit.